src/node/rpc/member_frontend.h

```diff
@@ -1332,16 +1332,56 @@ namespace ccf
         }
 
         const auto in = params.get<Propose::In>();
-        const auto proposal_id =
-          fmt::format("{:02x}", fmt::join(caller_identity.request_digest, ""));
 
-        Proposal proposal(in.script, in.parameter, caller_identity.member_id);
+        if (!consensus)
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "No consensus available.");
+        }
+
+        std::string proposal_id;
 
+        if (consensus->type() == ConsensusType::CFT)
+        {
+          auto root_at_read = ctx.tx.get_root_at_read_version();
+          if (!root_at_read.has_value())
+          {
+            return make_error(
+              HTTP_STATUS_INTERNAL_SERVER_ERROR,
+              ccf::errors::InternalError,
+              "Proposal failed to bind to state.");
+          }
+
+          // caller_identity.request_digest is set when getting the
+          // MemberSignatureAuthnIdentity identity. The proposal id is a digest
+          // of the root of the state tree at the read version and the request
+          // digest.
+          std::vector<uint8_t> acc(
+            root_at_read.value().h.begin(), root_at_read.value().h.end());
+          acc.insert(
+            acc.end(),
+            caller_identity.request_digest.begin(),
+            caller_identity.request_digest.end());
+          const crypto::Sha256Hash proposal_digest(acc);
+          proposal_id = proposal_digest.hex_str();
+        }
+        else
+        {
+          proposal_id = fmt::format(
+            "{:02x}", fmt::join(caller_identity.request_digest, ""));
+        }
+
+        Proposal proposal(in.script, in.parameter, caller_identity.member_id);
         auto proposals = ctx.tx.rw(this->network.proposals);
-        proposals->put(proposal_id, proposal);
+        if (!proposals->has(proposal_id))
```

Comment:
Is the `else` path here covered by a conflict when the tx is committed? It might be worth adding a short comment about this?

Reply:
No, it's not. What this does is if the proposal already exists, skip trying to create it, but still evaluate it in order to return an accurate status.

Reply:
AIUI, the only way the proposal ID could already exist (apart from hash collisions) is that two identical proposals were created at the same read version... which isn't possible because of this very line which takes a read dependency at this key on the proposal table (i.e. the second proposal creation would conflict).

Reply:
Yes, you're absolutely right, this can't even happen! I'll add a comment.